### ISSUE TITLE
C++ STL equivalent in Rust?

### ISSUE THREAD

**Issue by apoorv-2204** [rust-lang/rustlings#1382]
do we have c++ stl equivalent in rust?? Can we completely switch to Rust for competitive programming? or stick with c++

**Comment by cyril-marpaud** [comment]
[This link](https://www.geeksforgeeks.org/the-c-standard-template-library-stl/) describes the C++ STL.

Among others, it states that:

> Some of the key components of the STL include:
> - **Containers**: The STL provides a range of containers, such as vector, list, map, set, and stack, which can be used to store and manipulate data.
> - **Algorithms**: The STL provides a range of algorithms, such as sort, find, and binary_search, which can be used to manipulate data stored in containers.
> - **Iterators**: Iterators are objects that provide a way to traverse the elements of a container. The STL provides a range of iterators, such as forward_iterator, bidirectional_iterator, and random_access_iterator, that can be used with different types of containers.
> - **Function Objects**: Function objects, also known as functors, are objects that can be used as function arguments to algorithms. They provide a way to pass a function to an algorithm, allowing you to customize its behavior.
> - **Adapters**: Adapters are components that modify the behavior of other components in the STL. For example, the reverse_iterator adapter can be used to reverse the order of elements in a container.

In Rust:
- **Containers** are called [**Collections**](https://doc.rust-lang.org/stable/std/collections/)
- **Algorithms** are usually implemented on slices which enables you to call them on any type that implements the adequate trait. This is why you can call [sort](https://doc.rust-lang.org/std/primitive.slice.html#method.sort) or [binary_search](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) or any other [slice method](https://doc.rust-lang.org/std/primitive.slice.html) on an array or a `Vec`, for instance).
- [**Iterator**](https://doc.rust-lang.org/std/iter/trait.Iterator.html) is a trait and you can get an iterator out of any collection that implements the [IntoIterator trait](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html). The standard library implements it for you [for all standard collections](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#implementors) (`BTreeMap`, `HashMap`, `BinaryHeap`, `LinkedList`, `Option`, `BTreeSet`, `Vec`, `VecDecque`, `HashSet`, `Result`, `arrays`, `slices`...)
- **Function Objects** are called [*closures* or *function pointers*](https://doc.rust-lang.org/std/primitive.fn.html) and are described by 3 standard traits: [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html), [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) and [FnOnce](https://doc.rust-lang.org/std/ops/trait.FnOnce.html).
- **Adapters** are generally available as methods associated to the structures they refer to. See the [rev iterator method](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.rev), the [iterator reverse method](https://doc.rust-lang.org/std/primitive.slice.html#method.reverse) or also the [Reverse adapter](https://doc.rust-lang.org/std/cmp/struct.Reverse.html)...

In short, what you are looking for is already part of the [Rust Standard Library](https://doc.rust-lang.org/stable/std/).

**Comment by apoorv-2204** [comment]
> [This link](https://www.geeksforgeeks.org/the-c-standard-template-library-stl/) describes the C++ STL.
> 
> Among others, it states that:
> 
> > Some of the key components of the STL include:
> > 
> > * **Containers**: The STL provides a range of containers, such as vector, list, map, set, and stack, which can be used to store and manipulate data.
> > * **Algorithms**: The STL provides a range of algorithms, such as sort, find, and binary_search, which can be used to manipulate data stored in containers.
> > * **Iterators**: Iterators are objects that provide a way to traverse the elements of a container. The STL provides a range of iterators, such as forward_iterator, bidirectional_iterator, and random_access_iterator, that can be used with different types of containers.
> > * **Function Objects**: Function objects, also known as functors, are objects that can be used as function arguments to algorithms. They provide a way to pass a function to an algorithm, allowing you to customize its behavior.
> > * **Adapters**: Adapters are components that modify the behavior of other components in the STL. For example, the reverse_iterator adapter can be used to reverse the order of elements in a container.
> 
> In Rust:
> 
> * **Containers** are called [**Collections**](https://doc.rust-lang.org/stable/std/collections/)
> * **Algorithms** are usually implemented on slices which enables you to call them on any type that implements the adequate trait. This is why you can call [sort](https://doc.rust-lang.org/std/primitive.slice.html#method.sort) or [binary_search](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) or any other [slice method](https://doc.rust-lang.org/std/primitive.slice.html) on an array or a `Vec`, for instance).
> * [**Iterator**](https://doc.rust-lang.org/std/iter/trait.Iterator.html) is a trait and you can get an iterator out of any collection that implements the [IntoIterator trait](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html). The standard library implements it for you [for all standard collections](https://doc.rust-lang.org/std/iter/trait.IntoIterator.html#implementors) (`BTreeMap`, `HashMap`, `BinaryHeap`, `LinkedList`, `Option`, `BTreeSet`, `Vec`, `VecDecque`, `HashSet`, `Result`, `arrays`, `slices`...)
> * **Function Objects** are called [_closures_ or _function pointers_](https://doc.rust-lang.org/std/primitive.fn.html) and are described by 3 standard traits: [Fn](https://doc.rust-lang.org/std/ops/trait.Fn.html), [FnMut](https://doc.rust-lang.org/std/ops/trait.FnMut.html) and [FnOnce](https://doc.rust-lang.org/std/ops/trait.FnOnce.html).
> * **Adapters** are generally available as methods associated to the structures they refer to. See the [rev iterator method](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.rev), the [iterator reverse method](https://doc.rust-lang.org/std/primitive.slice.html#method.reverse) or also the [Reverse adapter](https://doc.rust-lang.org/std/cmp/struct.Reverse.html)...
> 
> In short, what you are looking for is already part of the [Rust Standard Library](https://doc.rust-lang.org/stable/std/).

So I can say goodbye to C++? as everything is there in Rust already part of the standard library ? we dont need any crates?. Beacuse in online judjes like codechef, leetcode, spoj, codeforces, they dont allow crates.

**Comment by cyril-marpaud** [comment]
Well, I really couldn't say, it depends on what you intend to do but a lot of people already use Rust for competitive programming. There even is an [advent_of_code crate](https://docs.rs/advent-of-code/latest/advent_of_code/) for instance, and you can find many other examples.

On the other hand, it's a little weird to me that you're asking for a "batteries-included competitive programming language" because it sounds like you're looking for a language that will give you an advantage over your competitors.

I personally believe that Rust and its ecosystem will help anyone be more productive but in the end, only you can answer that question by trying it. Be aware that the learning curve is a little steeper than for other languages, though.

**Comment by shadows-withal** [comment]
All that aside, this is a discussion forum for Rustlings. Your question might be better suited to one of the more general discussion forums, such as the [Users forum](https://users.rust-lang.org/). Closing this as off-topic.

**Comment by apoorv-2204** [comment]
> Well, I really couldn't say, it depends on what you intend to do but a lot of people already use Rust for competitive programming. There even is an [advent_of_code crate](https://docs.rs/advent-of-code/latest/advent_of_code/) for instance, and you can find many other examples.
> 
> On the other hand, it's a little weird to me that you're asking for a "batteries-included competitive programming language" because it sounds like you're looking for a language that will give you an advantage over your competitors.
> 
> I personally believe that Rust and its ecosystem will help anyone be more productive but in the end, only you can answer that question by trying it. Be aware that the learning curve is a little steeper than for other languages, though.

> sounds like you're looking for a language that will give you an advantage over your competitors.

I hope you are not oblivious to the original question.

**Comment by apoorv-2204** [comment]
> All that aside, this is a discussion forum for Rustlings. Your question might be better suited to one of the more general discussion forums, such as the [Users forum](https://users.rust-lang.org/). Closing this as off-topic.

Right, 
I really needed to have an insight. I just dont want to go back to C++.

**Comment by apoorv-2204** [comment]
Thank you everyone!
